### PR TITLE
feat(codewhisperer): factor in filedistance when fetching crossfile context

### PR DESCRIPTION
**Pull request by Will-ShaoHua**
### Before
- Crossfile file candidates are fetched by opened tabs without any sorting, e.g. files will be returned based on theirs orders in the IDE tabs (from the left to the right)

### After
- Crossfile file candidates fetched will be sorted by their file distance ascendingly. Files with smaller file distance will be prioritized than others.

## License

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

### CODE REVIEW

File: src/codewhisperer/util/supplementalContext/crossFileContextUtil.ts [hunk 1]
```diff
@@ -215,14 +216,28 @@ async function getCrossFileCandidates(editor: vscode.TextEditor): Promise<string
...
+                file: candidate,
+                fileDistance: getFileDistance(targetFile, candidate),
+            }
+        })
+        .sort((file1, file2) => {
+            return file1.fileDistance - file2.fileDistance
+        })
+        .map(fileToDistance => {
+            return fileToDistance.file
+        })
```
Comment (zixlin7) [inline src/codewhisperer/util/supplementalContext/crossFileContextUtil.ts:240]: nit: the mappings seem to be unnecessary? we can do `sort((file1, file2) => { return  getFileDistance(targetFile, file1) - getFileDistance(targetFile, file2) }`?

File: src/test/codewhisperer/util/crossFileContextUtil.test.ts [hunk 2]
```diff
@@ -49,6 +57,71 @@ describe('crossFileContextUtil', function () {
...
+    describe('getCrossFileCandidate', function () {
+        before(async function () {
+            this.timeout(60000)
+        })
+
+        beforeEach(async function () {
+            tempFolder = (await createTestWorkspaceFolder()).uri.fsPath
+        })
+
+        afterEach(async function () {
```
Comment (zixlin7) [inline src/test/codewhisperer/util/crossFileContextUtil.test.ts:69]: Justin mentioned this is not needed

**Review by justinmk3 — APPROVED** [review]

File: src/test/codewhisperer/util/crossFileContextUtil.test.ts [hunk 3]
```diff
@@ -49,6 +57,71 @@ describe('crossFileContextUtil', function () {
         })
     })
 
+    describe('getCrossFileCandidate', function () {
+        before(async function () {
+            this.timeout(60000)
```
Comment (justinmk3) [inline src/test/codewhisperer/util/crossFileContextUtil.test.ts:62]: Why does this test need so much time? We should avoid assigning giant timeouts for tests unless they really are doing a lot of work.

File: src/codewhisperer/util/supplementalContext/crossFileContextUtil.ts (hunk 1 above)
Comment (justinmk3) [reply src/codewhisperer/util/supplementalContext/crossFileContextUtil.ts:240]: Would calling `getFileDistance` in sort() would incur redundant calculations?

Comment (zixlin7) [reply src/codewhisperer/util/supplementalContext/crossFileContextUtil.ts:240]: I see, map makes sense if `getFileDistance` does heavy calculations

File: src/test/codewhisperer/util/crossFileContextUtil.test.ts (hunk 3 above)
Comment (Will-ShaoHua) [reply src/test/codewhisperer/util/crossFileContextUtil.test.ts:62]: gotcha, let me remove it, i referred to `visualizeStateMachine.test.ts` and thought this timeout is needed for tests manipulating editos or something lol

File: src/test/codewhisperer/util/crossFileContextUtil.test.ts (hunk 2 above)
Comment (Will-ShaoHua) [reply src/test/codewhisperer/util/crossFileContextUtil.test.ts:69]: wait, I thought this is needed, and I should add it in `beforeEach()` too as per the discussion here with Nick https://github.com/aws/aws-toolkit-vscode/pull/3668#discussion_r1270805968 @zixlin7 @justinmk3

Comment (zixlin7) [reply src/test/codewhisperer/util/crossFileContextUtil.test.ts:69]: Oh beforeEach is the one that has this done globally, it doesn't hurt to keep it in afterEach then

**Review by zixlin7 — APPROVED** [review]

**Pull request closed without merging**